File: Sources/Sentry/SentryDateUtil.m
#import "SentryDateUtil.h"
#import "SentrySwift.h"

NS_ASSUME_NONNULL_BEGIN

@interface SentryDateUtil ()

@property (nonatomic, strong) id<SentryCurrentDateProvider> currentDateProvider;

@end

@implementation SentryDateUtil

- (instancetype)initWithCurrentDateProvider:(id<SentryCurrentDateProvider>)currentDateProvider
{
    if (self = [super init]) {
        self.currentDateProvider = currentDateProvider;
    }
    return self;
}

- (BOOL)isInFuture:(NSDate *_Nullable)date
{
    if (date == nil)
        return NO;

    if (date == nil) {
        return NO;
    }
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wnullable-to-nonnull-conversion"
    NSComparisonResult result = [[self.currentDateProvider date] compare:date];
#pragma clang diagnostic pop
    return result == NSOrderedAscending;
}

+ (NSDate *_Nullable)getMaximumDate:(NSDate *_Nullable)first andOther:(NSDate *_Nullable)second
{
    if (first == nil && second == nil)
        return nil;
    if (first == nil)
        return second;
    if (second == nil)
        return first;

    if (second == nil) {
        return first;
    }
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wnullable-to-nonnull-conversion"
    NSComparisonResult result = [first compare:second];
#pragma clang diagnostic pop
    if (result == NSOrderedDescending) {
        return first;
    } else {
        return second;
    }
}

+ (long)millisecondsSince1970:(NSDate *)date
{
    return (long)([date timeIntervalSince1970] * 1000);
}

@end

NS_ASSUME_NONNULL_END

File: Sources/Swift/Helper/SentryDateUtil.swift
// swiftlint:disable missing_docs
import Foundation

@objc
@_spi(Private) public final class SentryDateUtil: NSObject {

    private let currentDateProvider: SentryCurrentDateProvider

    @objc(initWithCurrentDateProvider:)
    public init(currentDateProvider: SentryCurrentDateProvider) {
        self.currentDateProvider = currentDateProvider
        super.init()
    }

    @objc
    public func isInFuture(_ date: Date?) -> Bool {
        guard let date = date else {
            return false
        }

        let currentDate = currentDateProvider.date()
        return currentDate.compare(date) == .orderedAscending
    }

    @objc(getMaximumDate:andOther:)
    public static func getMaximumDate(_ first: Date?, andOther second: Date?) -> Date? {
        guard let first = first else {
            return second
        }
        guard let second = second else {
            return first
        }

        return first.compare(second) == .orderedDescending ? first : second
    }

    @objc
    public static func millisecondsSince1970(_ date: Date) -> Int {
        return Int(date.timeIntervalSince1970 * 1_000)
    }
}
// swiftlint:enable missing_docs
